### PR TITLE
[WC-120] - Implement Searched Course Page & [WC-41] - Implement Group/Online Group Landing Page

### PR DESCRIPTION
**Pull request by sikares**
# Description

[//]: # 'Please include a summary of the changes and the related issue. (You can write below without deleting this comment)'
This commit introduces a comprehensive search and filter functionality for the course listings on the WeCooked platform. The new features allow users to filter courses by cuisine types, dietary preferences, and a price range, enhancing user experience and navigation through course options.

### What

[//]: # 'Briefly describe this feature / fix / refactor (You can write below without deleting this comment)'
- Implemented a filtering system that includes:
  - Cuisine Type: Users can select from predefined cuisines.
  - Dietary Preferences: Users can filter courses based on various dietary requirements.
  - Price Range: A slider and input fields allow users to set minimum and maximum price limits for courses.
- The filtering options update dynamically based on the courses that match the selected criteria.
- Course cards display the filtered results, ensuring users can find relevant courses based on their search inputs.
- Calendar Feature:
  - Developed a user-friendly calendar interface displaying upcoming classes and events.
  - Users can view details about each event by clicking on the calendar dates.
- Advertisement Section:
  - Created a dedicated space for promotional ads to showcase relevant courses or partnerships.
  - The ad section can be dynamically updated with various promotional content.
- Upcoming Workshop Listing:
  - Implemented a dedicated section on the homepage for displaying upcoming workshops.
  - Users can see a list of future workshops with dates, times, and brief descriptions.

### Why

[//]: # 'Describe the reason why we need this feature / fix / refactor or refer to Jira Ticket (You can write below without deleting this comment)'
image
image
image
image
image

The addition of search and filter features is crucial for improving the user experience on the WeCooked platform, allowing users to easily navigate through various courses tailored to their preferences. This aligns with our goal of making the platform user-friendly and accessible, thereby increasing user engagement and satisfaction. 

The introduction of the calendar, advertisement section, and upcoming workshop listings is vital for enhancing user interaction on the WeCooked platform. These features allow users to stay informed about courses, events, and promotions, which contributes to a richer user experience and encourages participation in cooking workshops and classes.

These enhancements address the ongoing need for better visibility of events and promotions, aligning with our goal to make the platform more engaging and useful for users. This work responds to feedback collected during user testing and aims to fulfill requests for greater accessibility to event information.

## Type of change

Please check the relevent option with `x`.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (if needed)
- [x] My changes generate no new warnings
- [x] I have tested on my `Local Machine`

## Release Notes

- **New Features**
	- Introduced a new `SearchAndFilterPage` component for enhanced course searching and filtering based on user-defined criteria.
	- Added a `CustomCalendar` component for selecting date ranges.
	- Launched an `AdvertPage` component showcasing cooking techniques and ingredients.
	- Implemented a `GroupPage` component integrating various functionalities.
	- Added an `UpcomingWorkshopCard` component displaying a carousel of upcoming courses.

- **Improvements**
	- Updated multiple course entries across various data files, enhancing details such as titles, cuisines, prices, ratings, chefs, and dietary options.
	- Improved styling and layout for several components, including `SignUp`, `Login`, and `ForgetPassword`.
	- Enhanced the `SignUpChef` component with better error handling for photo uploads and improved button styling.

- **Bug Fixes**
	- Resolved minor syntax issues in components to ensure proper functionality.

### CODE REVIEW

File: app/client/data/ingredient.ts [hunk 1]
```diff
@@ -1,9 +1,7 @@
 export interface Ingredient {
   name: string
-  unit?: string // e.g., 'g', 'ml', 'pieces'
-  quantity?: number // Amount needed for the recipe
-  description?: string // Additional details or substitutes
-  category?: string // e.g., 'dairy', 'meat', 'pasta'
+  quantity?: SVGAnimatedNumber
```
[the remark this answers is not in the document]
Comment (sikares) [reply app/client/data/ingredient.ts:3]: Thank

File: app/client/video/search/[search]/page.tsx [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+import Footer from '@/app/client/components/Footer';
+import SearchAndFilterPage from '@/app/client/components/SearchAndFilterPage';
+
+export default function Page({ params }: { params: { search: string } }) {
+  return (
+    <div className="bg-gradient-to-b from-[#F0725C] to-[#FE3511] animate-fadeIn">
+      <NavbarAfter />
+      <SearchAndFilterPage params={params}/>
+      <Footer />
+    </div>
```
[the remark this answers is not in the document]
Comment (sikares) [reply app/client/video/search/[search]/page.tsx:12]: I will fix it.

File: app/client/components/FoodTypeCard.tsx [hunk 3]
```diff
@@ -67,24 +62,24 @@ export default function FoodTypeCard() {
...
             {foodtype.map((food, index) => (
               <Link
                 key={index}
-                href={`/client/video/search?type=${food.title.replace(
+                href={`/client/video/search/${(food.cuisine ?? '').replace(
                   /\s+/g,
                   '-'
                 )}`}
-                className="w-full pl-12"
+                className="w-full ml-12"
```
[the remark this answers is not in the document]
Comment (sikares) [reply app/client/components/FoodTypeCard.tsx:82]: Thank, I will fix it

File: app/client/components/SearchAndFilterPage.tsx [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+              </label>
+              <svg
+                className="absolute w-3.5 h-3.5 hidden peer-checked:block text-white pl-1"
+                xmlns="http://www.w3.org/2000/svg"
+                viewBox="0 0 22 22"
+                fill="none"
+                stroke="currentColor"
+                stroke-width="4"
+                stroke-linecap="round"
+                stroke-linejoin="round"
```
[the remark this answers is not in the document]
Comment (sikares) [reply app/client/components/SearchAndFilterPage.tsx:84]: I will fix it

File: app/client/components/SearchAndFilterPage.tsx [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+const CUISINE_TYPES = [
+  'Italian',
+  'French',
+  'Chinese',
+  'Thai',
+  'India',
+  'Japan',
+  'American',
+  'Mexican',
+]
```
[the remark this answers is not in the document]
Comment (sikares) [reply app/client/components/SearchAndFilterPage.tsx:30]: I will fix it

File: app/client/data/full-mock-data.ts [hunk 6]
```diff
@@ -0,0 +1,867 @@
...
+    course_id: 48,
+    title: 'Uma Parlour - Pure Vegetarian',
+    cuisine: 'India',
+    price: 300,
+    currency: 'THB',
+    rating: 4.0,
+    chef: 'Chef Sieng',
+    imageSrc: '/images/ramachandra.png',
+    chefImageUrl: '/images/chef.png',
+  },
```
[the remark this answers is not in the document]
Comment (sikares) [reply app/client/data/full-mock-data.ts:590]: I will fix it

File: app/client/components/AdvertPage.tsx [hunk 7]
```diff
@@ -0,0 +1,77 @@
+import React from 'react'
+import Image from 'next/image'
+
+export default function AdvertPage() {
+  return (
+    <div className="h-[800px] relative w-full flex flex-col items-center">
```
[the remark this answers is not in the document]
Comment (sikares) [reply app/client/components/AdvertPage.tsx:6]: I will fix it

File: app/client/components/AdvertPage.tsx [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+          >
+            Authentic
+          </button>
+        </div>
+        <div className="flex">
+          <button className="px-7 py-2 bg-gradient-to-b from-[#F0725C] to-[#FE3511] text-white rounded-full font-semibold transition-all duration-300 transform hover:scale-105 hover:shadow-lg">
+            Let&apos;s Cook
+          </button>
+        </div>
+      </div>
```
[the remark this answers is not in the document]
Comment (sikares) [reply app/client/components/AdvertPage.tsx:65]: I will fix it

File: app/client/components/CalendarPage.tsx [hunk 9]
```diff
@@ -0,0 +1,127 @@
+'use client'
+import React, { useState, useRef, useEffect } from 'react'
+import Image from 'next/image'
+import { MagnifyingGlassIcon, CalendarIcon } from '@heroicons/react/24/solid'
+import CustomCalendar from './CustomCalemdar'
```
[the remark this answers is not in the document]
Comment (sikares) [reply app/client/components/CalendarPage.tsx:5]: I will fix it

File: app/client/components/CustomCalendar.tsx [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+    if (!value.startDate || (value.startDate && value.endDate)) {
+      onChange({ startDate: date, endDate: null })
+    } else {
+      if (date < value.startDate) {
+        onChange({ startDate: date, endDate: value.startDate })
+      } else {
+        onChange({ startDate: value.startDate, endDate: date })
+      }
+    }
+  }
```
[the remark this answers is not in the document]
Comment (sikares) [reply app/client/components/CustomCalendar.tsx:88]: I will fix it later

File: app/client/data/upcoming-course.ts [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+  {
+    course_id: 0,
+    title: 'Authentic Italian - Pasta Masterclass',
+    cuisine: 'Italian',
+    price: 500,
+    currency: 'THB',
+    rating: 4.5,
+    chef: 'Chef Lorenzo',
+    imageSrc: '/images/ramachandra.png',
+    chefImageUrl: '/images/chef.png',
```
[the remark this answers is not in the document]
Comment (sikares) [reply app/client/data/upcoming-course.ts:13]: I think it not necessary

File: app/client/components/CustomCalendar.tsx [hunk 12]
```diff
@@ -0,0 +1,188 @@
+import React, { useState } from 'react'
+import { ChevronLeftIcon, ChevronRightIcon } from '@heroicons/react/16/solid'
+
+interface CustomCalendarProps {
+  value: { startDate: Date | null; endDate: Date | null }
+  onChange: (value: { startDate: Date | null; endDate: Date | null }) => void
+}
```
[the remark this answers is not in the document]
Comment (sikares) [reply app/client/components/CustomCalendar.tsx:7]: I think it not necessary

File: app/client/components/CustomCalendar.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (sikares) [reply app/client/components/CustomCalendar.tsx:88]: I think it not necessary

File: app/client/components/CustomCalendar.tsx [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+
+const CustomCalendar: React.FC<CustomCalendarProps> = ({ value, onChange }) => {
+  const [currentDate, setCurrentDate] = useState(new Date())
+  const [nextMonthDate, setNextMonthDate] = useState(
+    new Date(new Date().setMonth(currentDate.getMonth() + 1))
+  )
```
[the remark this answers is not in the document]
Comment (sikares) [reply app/client/components/CustomCalendar.tsx:13]: I think it not necessary

File: app/client/components/CustomCalendar.tsx [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+                {day}
+              </div>
+            ))}
+            {renderCalendar(nextMonthDate)}
+          </div>
+        </div>
+      </div>
+    </div>
+  )
+}
```
[the remark this answers is not in the document]
Comment (sikares) [reply app/client/components/CustomCalendar.tsx:186]: I think it not necessary

File: app/client/components/CustomCalendar.tsx [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+                : ''
+            }`}
+        >
+          {day}
+        </div>
+      )
+    }
+
+    return days
+  }
```
[the remark this answers is not in the document]
Comment (sikares) [reply app/client/components/CustomCalendar.tsx:128]: I think it not necessary

File: app/client/components/UpcomingWorkshopCard.tsx [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+              className="flex transition-transform duration-300 ease-in-out"
+              style={{
+                transform: `translateX(-${startIndex * (100 / cardsToShow)}%)`,
+              }}
+            >
+              {courses.map((course, index) => (
+                <div key={index} className="w-full pl-12">
+                  <CourseCard {...course} />
+                </div>
+              ))}
```
[the remark this answers is not in the document]
Comment (sikares) [reply app/client/components/UpcomingWorkshopCard.tsx:64]: I think it not necessary

**Comment by tong788** [comment]
image
Can you change it to light color as figma?
image

File: app/client/components/CalendarPage.tsx [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+                <MagnifyingGlassIcon className="h-5 w-5 text-gray-400 absolute left-3 bottom-3" />
+              </div>
+              <div className="space-y-2 relative">
+                <p className="mt-6">Select Date Range</p>
+                <input
+                  ref={inputRef}
+                  type="text"
+                  readOnly
+                  placeholder="Select Date Range"
+                  value={formatDateRange()}
```
[the remark this answers is not in the document]
Comment (sikares) [reply app/client/components/CalendarPage.tsx:96]: I will fix it.

**Review by tong788 — APPROVED** [review]

**Pull request merged**